frontend/apps/reader/modules/readersearch.lua

```diff
@@ -145,6 +146,17 @@ function ReaderSearch:onShowSearchDialog(text, direction)
                 -- Don't add result pages to location ("Go back") stack
                 neglect_current_location = true
             end
+            if res==nil or (current_page~=nil and valid_link==nil) then
```

Comment:
This is a bit hard to understand - and may give doubts :) valid_link and current_page are not defined on PDFs (when `if self.ui.document.info.has_pages`).
Might be clearer if you add another bool variable `local has_results = false` and do simpler checks in the appropriate branches.

Reply:
> valid_link and current_page are not defined on PDFs

Yes, that's the fact but...
I've tested it with several pdfs and it does work as designed.
Will you be so kind to check it with a real ebook?

Reply:
I won't be that kind :) as the issue is not about it working or not, just about the code being clearer :)
I'd rather see something like this:
```diff
--- a/frontend/apps/reader/modules/readersearch.lua
+++ b/frontend/apps/reader/modules/readersearch.lua
@@ -73,11 +73,13 @@ function ReaderSearch:onShowSearchDialog(text, direction)
     local current_page
     local do_search = function(search_func, _text, param)
         return function()
+            local has_result = false
             local res = search_func(self, _text, param)
             if res then
                 if self.ui.document.info.has_pages then
                     self.ui.link:onGotoLink({page = res.page - 1}, neglect_current_location)
                     self.view.highlight.temp[res.page] = res
+                    has_result = true
                 else
                     -- Was previously just:
                     --   self.ui.link:onGotoLink(res[1].start, neglect_current_location)
@@ -140,10 +142,14 @@ function ReaderSearch:onShowSearchDialog(text, direction)
                     end
                     if valid_link then
                         self.ui.link:onGotoLink({xpointer=valid_link}, neglect_current_location)
+                        has_result = true
                     end
                 end
                 -- Don't add result pages to location ("Go back") stack
                 neglect_current_location = true
+                if not has_result then
+                    -- do notification stuff
+                end
             end
         end
     end
```
than having to wonder what cases `if res==nil or (current_page~=nil and valid_link==nil) then` checks for (because I still wonder :) I'm not sure the above `has_result` stuff is enough, or if you thought about other cases depending on current_page meaning that a first search has already been done, etc... if yes, this could be made  clearer with `has_result=true` in the appropriate places).

Reply:
Thank you for your questions, now I can explain why my code works for pdf.
For initial search when nothing found - it's clear, `res=nil`.

More complicated is case when we are, say, on the last page with search results and press `▷`. We want to see notification `No results forward`.

The point is that in that case `findText` returns different results for epub and for pdf!
For `self.ui.document.info.has_page` it returns `res=nil` and we need no more checks.

Regading proposition of adding several `has_result=true`. They all should be under `if` conditions for `current_page` and `valid_link`.
I've just traced the logic and compacted it into one expression.
The code is short and it does work.

Reply:
It's still hard to understand. For example:
`current_page~=nil and valid_link==nil` are you sure the `current_page~=nil ` check is necessary ?
Isn't just `if res==nil or valid_link==nil then` enough ?

> The code is short and it does work.

The fact we're having this discussion may show these 2 conditions are not enough :)

Reply:
> are you sure the `current_page~=nil ` check is necessary ?

Yes it is, without it we shall get notification for every search for `self.ui.document.info.has_page` as `valid_link=nil` always for them.

Reply:
So, `current_page` serves as an indirect far fetched equivalent of `not self.ui.document.info.has_pages` (and not even its main meaning to distinguish initial vs subsequence searches) :/

Make it easy for future readers of that bit of code (that may be me or you :):
- there's no reason to move `valid_link` outside the else where it matters
- add a natural language `has_result` so this reads naturally, without having to wonder how/why/when a convoluted `if res==nil or (current_page~=nil and valid_link==nil) then` will trigger, like we're doing for the last 5 hours :)